ext/console/02_jsx.js

```diff
@@ -0,0 +1,431 @@
+    }
+    if (vnode.type == SymbolFor("react.fragment")) return "Fragment";
+    if (vnode.type == SymbolFor("react.suspense")) return "Suspense";
+    if (vnode.type == SymbolFor("react.strict_mode")) return "StrictMode";
+    if (vnode.type == SymbolFor("react.profiler")) return "Profiler";
+
+    if (
+      vnode.type !== null &&
+      typeof vnode.type === "object" &&
+      typeof vnode.type.$$typeof === "symbol"
+    ) {
+      const $$typeof = vnode.type.$$typeof;
+      if ($$typeof === SymbolFor("react.provider")) {
+        const suffix = vnode.type.displayName
+          ? "." + vnode.type.displayName
+          : "";
+        return `Provider${suffix}`;
+      } else if ($$typeof === SymbolFor("react.context")) {
+        const suffix = vnode.type.displayName
+          ? "." + vnode.type.displayName
+          : "";
+        return `Consumer${suffix}`;
+      } else if (
+        $$typeof === SymbolFor("react.memo") &&
+        typeof vnode.type.type === "function"
+      ) {
+        const fnName = vnode.type.type.displayName || vnode.type.type.name ||
+          "";
+        return `Memo${fnName ? `(${fnName})` : ""}`;
+      } else if ($$typeof === SymbolFor("react.forward_ref")) {
+        const name = vnode.type.render.displayName ||
+          vnode.type.render.name ||
+          "Anonymous";
+        return `ForwardRef(${name})`;
+      }
+    }
+
+    return "Unknown";
+  },
+  getTextIfTextNode() {
+    // React doesn't have proper text elements afaik
+    return null;
+  },
+  isFragment(vnode) {
+    return vnode.type === SymbolFor("react.fragment");
+  },
+};
+
+/**
+ * Preact adapter to serialize JSX to string
+ * @type {import("ext:deno_console/jsx").JSXSerializeAdapter}
+ */
+export const preactAdapter = {
+  getName(vnode) {
+    if (typeof vnode.type === "string") return vnode.type;
+    if (typeof vnode.type === "function") {
+      const name = vnode.type.displayName || vnode.type.name || "Anonymous";
+      // TODO: Fragments can only be detected by importing them, this is
+      // a hacky workaround
+      if (name === "k") return "Fragment";
+
+      if ("contextType" in vnode.type) {
+        const ct = vnode.type.contextType;
+        const name = vnode.type === ct.Consumer ? "Consumer" : "Provider";
+        const suffix = ct.displayName ? `.${ct.displayName}` : "";
+        return name + suffix;
+      }
+      return name;
+    }
+    return "Unknown";
+  },
+  getTextIfTextNode(vnode) {
+    return vnode.type === null ? String(vnode.props.data) : null;
+  },
+  isFragment(vnode) {
+    // TODO: Fragments can only be detected by importing them, this is
+    // a hacky workaround
+    return typeof vnode.type === "function" && vnode.type.name === "k";
+  },
+  isValidElement(value) {
+    return value !== null && typeof value === "object";
+  },
+};
+
+function isVNode(x) {
+  return (
+    x !== null &&
+    typeof x === "object" &&
+    "type" in x &&
+    "props" in x &&
+    "key" in x
+  );
+}
```

Comment:
One thing worth to check (can do on Monday) is to inline the framework's `isValideElement` calls into the adapter and use that for JSX detection as well as vnode detection. It's likely a more foolproof way, especially for the case when fake vnodes are injected.

Should be a pretty minor change. I'll add that on Monday.